pkg/controller/direct/apikeys/apikeyskey/apikeyskey_controller.go

```
@@ -0,0 +1,296 @@
+	// 	opts = append(opts, option.WithEndpoint(m.config.Endpoint))
+	// }
+
+	gcpClient, err := apikeys.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building apikeys client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	gcp, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Just fetch this object?
+	obj := &v1alpha1.APIKeysKey{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	projectID := obj.Spec.ProjectRef.External
+	if projectID == "" {
+		return nil, fmt.Errorf("unable to determine project")
+	}
+
+	// TODO: Use name or request resourceID to be set on create?
+	keyID := ValueOf(obj.Spec.ResourceID)
+	if keyID == "" {
+		return nil, fmt.Errorf("unable to determine resourceID")
+	}
+
+	location := "global"
+
+	return &adapter{
+		projectID: projectID,
+		location:  location,
+		keyID:     keyID,
+		desired:   obj,
+		gcp:       gcp,
+	}, nil
+}
+
+func ValueOf[T any](p *T) T {
+	var v T
+	if p != nil {
+		v = *p
+	}
+	return v
+}
+
+func (a *adapter) Find(ctx context.Context) (bool, error) {
+	if a.keyID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	key, err := a.gcp.GetKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			klog.Warningf("key was not found: %v", err)
+			return false, nil
+		}
+		return false, err
+	}
+
+	u := &v1alpha1.APIKeysKey{}
+	if err := keyMapping.Map(key, u); err != nil {
+		return false, err
+	}
+	a.actual = u
+
+	return true, nil
+}
+
+func (a *adapter) Delete(ctx context.Context) error {
+	// TODO: Delete via status selfLink?
+	req := &pb.DeleteKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	op, err := a.gcp.DeleteKey(ctx, req)
+	if err != nil {
+		return fmt.Errorf("deleting key: %w", err)
```

Should we return an error if the object matching the key has already been deleted?

---

Good call.  I don't think we should return an error, but I do now return a bool to tell the difference between "I really deleted this" and "The object was already gone"